Migration: Upload migration files from profile settings and start processing

tclaus opened the pull request:
Currently a import from Settings, while a user already exists. 

ToDo:

- [x] Photos upload and processing compressed file
- [x] Photos processing
- [x] Rake file adaptions
- [x] Use translation for texts

tclaus committed:
Upload migration files from profile settings and start processing

Moving Texts to localization file

This is a rebase on latest dev and a squash of the fixing commits

- Updates (most) texts to goobs proposal
- Moved texts to localization file
- Fixed file-upload buttons
- Added photos import to rake file
- No submit if no photo is selected

On profile file upload the same fields in the user table are used than for generated exporting files.
This conflicts if any export files exists and user wants to import something

diaspora-code-review commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,53 @@
+          %p
+            If the pod of the account you are importing is offline, your contacts and posts will still be imported,
+            but your contacts will receive a notification as if you started sharing with them.
+          = form_for current_user, url: edit_user_path, html: {method: :put, multipart: true, class: "form-horizontal"} do |f|
```
Line is too long. [126/120]

diaspora-code-review commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,53 @@
+              .clearfix
+                .btn.btn-default{data: {dismiss: "modal"}}
+                  Cancel
+                = f.submit 'Import and merge', class: "btn btn-primary.pull-right", id: "change_email_preferences"
```
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,53 @@
+                .btn.btn-default{data: {dismiss: "modal"}}
+                  Cancel
+                = f.submit 'Import and merge', class: "btn btn-primary.pull-right", id: "change_email_preferences"
+                  
```
Line contains trailing whitespace

diaspora-code-review commented on app/workers/import_profile.rb:
```diff
@@ -0,0 +1,57 @@
+      end
+    end
+
+private
```
- Layout/AccessModifierIndentation: Indent access modifiers like `private`.
- Layout/EmptyLinesAroundAccessModifier: Keep a blank line before and after `private`.

diaspora-code-review commented on app/workers/import_profile.rb:
```diff
@@ -0,0 +1,57 @@
+        photo.remote_photo_name = nil # migration sets this attribute, but file needs to be procesed
+        photo.unprocessed_image = photo_filename
+        photo.save(touch: false)
+
```
Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body end.

diaspora-code-review commented on app/workers/import_profile.rb:
```diff
@@ -0,0 +1,57 @@
+    def unzip_photos_file(photo_file_path)
+      system("unzip", photo_file_path)
+    end
+
```
Layout/EmptyLinesAroundClassBody: Extra empty line detected at class body end.

diaspora-code-review commented on app/uploaders/exported_photos.rb:
```diff
@@ -9,7 +9,14 @@ def store_dir
+
   def filename
-    "#{model.username}_photos_#{secure_token}.zip" if original_filename.present?
+    if original_filename.present?
```
Style/GuardClause: Use a guard clause (`return unless original_filename.present?`) instead of wrapping the code inside a conditional expression.

diaspora-code-review commented on app/uploaders/exported_user.rb:
```diff
@@ -10,10 +10,13 @@ def store_dir
 
   def filename
-    "#{model.username}_diaspora_data_#{secure_token}.json.gz" if original_filename.present?
+    if original_filename.present?
```
Style/GuardClause: Use a guard clause (`return unless original_filename.present?`) instead of wrapping the code inside a conditional expression.

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -96,20 +96,22 @@ def processed?
-  def url(name = nil)
-    if remote_photo_path
-      name = name.to_s + '_' if name
+  def url(name=nil)
```
Metrics/AbcSize: Assignment Branch Condition size for url is too high. [<2, 19, 7> 20.35/20]

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -96,20 +96,22 @@ def processed?
+    # During migration in bad cases this might be happen
+    # If this happens, stream loading stops. Better dont show photos as stop loading stream
+    if remote_photo_path.present? && remote_photo_name.present?
+      name = "#{name.to_s}_" if name
```
Lint/RedundantStringCoercion: Redundant use of `Object#to_s` in interpolation.

goobertron commented:
I'd like chance to review the text (including code comments) at some point, as Diaspora's (sort of) 'chief proof-reader'. I'll try to do this soon.

goobertron commented on app/controllers/users_controller.rb:
```diff
@@ -235,6 +239,22 @@ def change_email(user_data)
+    @user.export = user_data[:export] if user_data[:export]
+    @user.exported_photos_file = user_data[:exported_photos_file] if user_data[:exported_photos_file]
+    if @user.save
+      flash.now[:notice] = "A profile migration is scheduled"
```
Assuming this includes data and photos, I suggest using 'account' rather than 'profile'. How about either:
- Your account migration has been scheduled
or
- Your account migration is in progress

goobertron commented on app/controllers/users_controller.rb:
```diff
@@ -235,6 +239,22 @@ def change_email(user_data)
+    if @user.save
+      flash.now[:notice] = "A profile migration is scheduled"
+    else
+      flash.now[:error] = "An error occured scheduling a migration: #{@user.errors.full_messages}"
```
It looks as though a specific error message is appended to this warning, so I suggest:
- The following error occurred in scheduling your account migration:
or
- Your account migration could not be scheduled for the following reason:
(note: two Rs in occurred)

goobertron commented on app/models/account_migration.rb:
```diff
@@ -29,6 +29,7 @@ def sender
   # executes a migration plan according to this AccountMigration object
```
Could this be clearer? Or is it clear to any coder?

goobertron commented on app/models/photo.rb:
```diff
@@ -48,25 +48,25 @@ class Photo < ApplicationRecord
+    message = StatusMessage.find_by(guid: status_message_guid)
     return unless status_message_guid && message && diaspora_handle != message.diaspora_handle
 
     errors.add(:base, "Photo must have the same owner as status message")
```
If this is text in the UI, I'll need to know context in order to be able to polish this.

tclaus replied:
This is internally only. I think, a photo must have a reference to a post ('status message' internally) to be valid. This text goes out to an error log.

goobertron replied:
Got it. That is probably clear enough to a coder or admin.

goobertron commented on app/models/photo.rb:
```diff
@@ -96,20 +96,22 @@ def processed?
-    if remote_photo_path
-      name = name.to_s + '_' if name
+  def url(name=nil)
+    # During migration in bad cases this might be happen
```
Again, these two lines could be clearer, but context would be helpful.

tclaus replied:
A message to a coder. The if clause the next line was changed, because in rare cases a photo can not be processed. Then the stream will not load. Diaspora stops (crashes). Would be nice to use clear words for this.

goobertron replied:
How about 'If a photo cannot be processed during migration, the migration will fail and the stream will stop loading. If this happens, don't show the photos' ???

goobertron commented on app/models/user.rb:
```diff
@@ -312,6 +312,9 @@ def like_for(target)
   ######### Data export ##################
   mount_uploader :export, ExportedUser
 
+  ######### Photos export ##################
```
'Photo export' is better that Photos (plural). I can't really explain why!

goobertron commented on app/models/user.rb:
```diff
@@ -325,29 +328,26 @@ def perform_export!
-    logger.error "Unexpected error while exporting user '#{username}': #{error.class}: #{error.message}\n" \
-                 "#{error.backtrace.first(15).join("\n")}"
+  rescue StandardError => e
+    logger.error "Unexpected error while exporting user '#{username}': #{e.class}: #{e.message}\n" \
```
- Unexpected error while exporting data for '#{username}':
This mirrors the error message for photos.

goobertron commented on app/views/users/_edit.haml:
```diff
@@ -193,8 +193,11 @@
-      .col-md-6.account-data
+      .col-md-12
         %h3= t(".export_data")
+        %p If you want to migrate this account to another pod, start by exporting your data using the buttons below.
```
... start by exporting your data and photos ...

goobertron commented on app/views/users/_edit.haml:
```diff
@@ -224,7 +228,24 @@
+          Import another account into
+          = current_user.diaspora_handle
+        %p
+          diaspora* allows you to merge an account into another, even if the accounts are in two different pods.
```
I'll have to come back to this text block. Do remind me.

goobertron replied:
Suggested some modified texts here, for discussion: https://discourse.diasporafoundation.org/t/backup-and-restore-account-migration/544/173

goobertron commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,55 @@
+                %label Select the data archive:
+                = f.file_field :export, accept: "application/json, application/zip, application/gzip"
+              .col-md-6
+                %label Select the photos archive:
```
'photo archive'

goobertron commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,55 @@
+                %label Select the photos archive:
+                = f.file_field :exported_photos_file, accept: "application/zip"
+            %h4
+              The original account will be imported
```
You data will be imported %strong and the original account deleted.

goobertron replied:
This needs to be transferred to the i18n file in `config`. Also, I still think it's better not to repeat 'the original account' in this line, so change the first one to 'Your data will be imported'.

And maybe add 'This cannot be undone.' after this sentence?

goobertron reviewed (commented):
I've reviewed most of the strings and made suggested changes to some of them. I think I've seen everything, but it's not easy with the text in code files rather than in a separate config file. I'll check again once the text has been extracted.

There are a few comments where I'm not sure exactly what they are supposed to communicate, so I've asked for clarification.

I'll have to think about the two large text blocks at a later date, so I'll come back to them.

Thanks!

diaspora-code-review commented on app/views/users/_edit.haml:
```diff
@@ -193,8 +193,11 @@
-      .col-md-6.account-data
+      .col-md-12
         %h3= t(".export_data")
+        %p If you want to migrate this account to another pod, start by exporting your data and photos using the buttons below.
```
Line is too long. [127/120]

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+const profileFileBtn = document.getElementById('profile-file-btn');
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+const profileFileBtn = document.getElementById('profile-file-btn');
+
+const profileFileChosen = document.getElementById('profile-file-chosen');
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+
+const profileFileChosen = document.getElementById('profile-file-chosen');
+
+profileFileBtn.addEventListener('change', function(){
```
- Strings must use doublequote.
- Missing space before opening brace.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+const profileFileChosen = document.getElementById('profile-file-chosen');
+
+profileFileBtn.addEventListener('change', function(){
+  profileFileChosen.textContent = this.files[0].name
```
Missing semicolon.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+
+profileFileBtn.addEventListener('change', function(){
+  profileFileChosen.textContent = this.files[0].name
+})
```
Missing semicolon.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+  profileFileChosen.textContent = this.files[0].name
+})
+
+const photosFileBtn = document.getElementById('photos-file-btn');
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+
+const photosFileBtn = document.getElementById('photos-file-btn');
+
+const photosFileChosen = document.getElementById('photos-file-chosen');
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+
+const photosFileChosen = document.getElementById('photos-file-chosen');
+
+photosFileBtn.addEventListener('change', function(){
```
- Strings must use doublequote.
- Missing space before opening brace.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+const photosFileChosen = document.getElementById('photos-file-chosen');
+
+photosFileBtn.addEventListener('change', function(){
+  photosFileChosen.textContent = this.files[0].name
```
Missing semicolon.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,15 @@
+
+photosFileBtn.addEventListener('change', function(){
+  photosFileChosen.textContent = this.files[0].name
+})
```
Missing semicolon.

diaspora-code-review commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,57 @@
+                .upload-btn-wrapper
+                  .btn.btn-primary
+                    Select profile archive
+                  = f.file_field :export, accept: "application/json, application/zip, application/gzip",
```
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,57 @@
+                .upload-btn-wrapper
+                  .btn.btn-primary
+                    Select photo archive
+                  = f.file_field :exported_photos_file, accept: "application/zip",
```
- Layout/SpaceAfterColon: Space missing after colon.
- Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/assets/stylesheets/forms.scss:
```diff
@@ -18,6 +18,20 @@ textarea {
+  display: inline-block;
+}
+
+.upload-btn-wrapper input[type=file] {
```
- Merge rule `.upload-btn-wrapper input[type=file]` with rule on line 21
- Avoid qualifying attribute selectors with an element.

diaspora-code-review commented on app/assets/stylesheets/forms.scss:
```diff
@@ -18,6 +18,20 @@ textarea {
 // scss-lint:enable QualifyingElement
 
+.upload-btn-wrapper {
+  position: relative;
```
Properties should be ordered display, overflow, position

diaspora-code-review commented on app/assets/stylesheets/forms.scss:
```diff
@@ -18,6 +18,20 @@ textarea {
+
+.upload-btn-wrapper input[type=file] {
+  font-size: 100px;
+  position: absolute;
```
Properties should be ordered font-size, left, opacity, position, top

diaspora-code-review commented on lib/archive_importer.rb:
```diff
@@ -36,7 +36,8 @@ def create_user(attr)
     )
-    self.user = User.build(data)
+    self.user = User.find_or_build(data)
+    self.user.getting_started = false
```
Style/RedundantSelf: Redundant `self` detected.

diaspora-code-review commented on app/assets/stylesheets/forms.scss:
```diff
@@ -18,6 +18,20 @@ textarea {
+  overflow: hidden;
+  position: relative;
+
+  input[type=file] {
```
Avoid qualifying attribute selectors with an element.

tclaus commented:
Current status of the import dialog is: 

![Bildschirmfoto 2021-09-08 um 23 14 08](https://user-images.githubusercontent.com/501326/132979353-96f62406-370b-4e99-94fa-18616e0e61f4.png)

Rake file allows import of photo archives:
![Bildschirmfoto 2021-09-12 um 10 03 16](https://user-images.githubusercontent.com/501326/132979361-81127f91-b1f7-4f1a-832e-9c0edebe5991.png)

goobertron commented on config/locales/diaspora/en.yml:
```diff
@@ -1327,7 +1334,21 @@ en:
+      list3: "All your aspects and contacts from that account will be added to this account"
+      list4: "All posts and comments from that account will be added to this account"
+      closing_text: "The imported account and all its linked data will be deleted from the previous pod.<br>
+                     Your contacts do not need to do anything; they will automatically be connected to this account.<br>
```
I would not have the rest of this text in bold – only 'The imported account and all its linked data will be deleted from the previous pod.'

I also suggest separating this text block into three strings, one for each paragraph, rather than using <br>, which I think is not favoured.

goobertron reviewed (commented):
Thanks. I can only comment on the wording (and the fact that all the text needs to be in ` config/locales/diaspora/en.yml`.

Fla has said in [this Discourse comment](https://discourse.diasporafoundation.org/t/backup-and-restore-account-migration/544/175) that the text is not that accurate, so we might have to rewrite some of it. But with the two minor changes I've just suggested, I think it reads well enough and is clear enough, even if some of it is not yet accurate!

Flaburgan commented:
Hey, could you please make the rake test to fail when it doesn't find the archive?

diaspora-code-review commented on app/views/users/_import_account_modal.haml:
```diff
@@ -0,0 +1,62 @@
+              .clearfix
+                .btn.btn-default{data: {dismiss: "modal"}}
+                  = t("users.import_modal.cancel")
+                = f.submit t("users.import_modal.start_import"), class: "btn btn-primary.pull-right", id: "change_email_preferences"
```
Line is too long. [132/120]

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+
+profileFileButton.addEventListener("change", function() {
+  profileFileChosen.textContent = this.files[0].name;
+  checkProfileUploadButton();
```
'checkProfileUploadButton' was used before it was defined

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+
+photosFileButton.addEventListener("change", function() {
+  photosFileChosen.textContent = this.files[0].name;
+  checkProfileUploadButton();
```
'checkProfileUploadButton' was used before it was defined

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+});
+
+const checkProfileUploadButton = function() {
+  let photo_files = $("#profile-file-btn")[0].files;
```
Identifier 'photo_files' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+
+const checkProfileUploadButton = function() {
+  let photo_files = $("#profile-file-btn")[0].files;
+  let profile_files = $("#photos-file-btn")[0].files;
```
Identifier 'profile_files' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+  let photo_files = $("#profile-file-btn")[0].files;
+  let profile_files = $("#photos-file-btn")[0].files;
+
+  if ((photo_files.size + profile_files.size) === 0)
```
- Opening curly brace does not appear on the same line as controlling statement.
- Identifier 'photo_files' is not in camel case.
- Identifier 'profile_files' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+
+  if ((photo_files.size + profile_files.size) === 0)
+  {
+    $("#upload_profile_files").attr('disabled', 'disabled');
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+  {
+    $("#upload_profile_files").attr('disabled', 'disabled');
+  } else {
+    $("#upload_profile_files").removeAttr('disabled');
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/archive-uploader.js:
```diff
@@ -0,0 +1,30 @@
+    $("#upload_profile_files").removeAttr('disabled');
+  }
+
+}
```
- Block must not be padded by blank lines.
- Missing semicolon.

Flaburgan commented:
Hey @tclaus could you please extract the rake refactor and the photos import (basically everything only linked to the backend) in a smaller pull request? This would allow me to start doing more tests even if the UI isn't ready yet

Flaburgan commented:
So, I tried the UI locally and it didn't work, the validation took too much time and the HTTP connection "has been reinitialized" (says Firefox). I think we need to split this again. Maybe we could have a different setting page so it's actually easier to find (and to link to) for the user?
So we would add an "Export / Import" item in the settings menu. There, we have more space to actually put the import buttons, and allow the user to come back later even for the validation step.

tclaus commented:
> So, I tried the UI locally and it didn't work, the validation took too much time and the HTTP connection "has been reinitialized" (says Firefox). I think we need to split this again. Maybe we could have a different setting page so it's actually easier to find (and to link to) for the user?
> So we would add an "Export / Import" item in the settings menu. There, we have more space to actually put the import buttons, and allow the user to come back later even for the validation step.

An extra menu is nice. 
There is no "wait" for the validation. Instantly after uploading the files, a job should start with validation and processing. 
Did you test it locally.. without a running sidekiq for background processing? I have experienced something similar.

Flaburgan commented:
> There is no "wait" for the validation. Instantly after uploading the files, a job should start with validation and processing.

But what happens for you when you click the button in the modal then? For me, it does launch the validation, but the modal stays open waiting for it to finish, and with real data, that takes too much time and the http connection expires.

SuperTux88 commented:
If you run diaspora locally in dev-mode there is no sidekiq by default (the jobs happen in the request then, which works for most things needed for dev-mode), for certain things like migration (takes too long for being handled in request) and also federation with other local pods (with only 1 unicorn worker that is already busy handling the federation job it doesn't have another worker free to handle a webfinger request from a second pod, so it deadlocks), you need a sidekiq, so you need to enable that in the `diaspora.toml`.

Flaburgan commented:
Ok, I was able to enable sidekiq by setting `single_process_mode = false`. Thank you.
Indeed, the modal then close, but when the import fails, there is no message in the settings page. The button to import should disappear and there should be a feedback to the user.
After submitting the modal: "Processing archive XXX" 
If the validation failed: "Archive is invalid, ERROR_MESSAGE"
If the import failed: "Import failed, ERROR_MESSAGE, please warn your podmin"
If the import succeed: "Archive XXX imported successfully" and the button to import should reappear.

If you like the idea of having all this on a separated page I can create it.

diaspora-code-review commented on app/controllers/users_controller.rb:
```diff
@@ -122,6 +122,10 @@ def confirm_email
     redirect_to edit_user_path
   end
 
+  def has_import_parameter?(import_parameters)
```
Naming/PredicateName: Rename `has_import_parameter?` to `import_parameter?`.

diaspora-code-review commented on app/controllers/users_controller.rb:
```diff
@@ -229,15 +233,30 @@ def change_email(user_data)
+
+  def copy_import_files(user_data)
+    {
+            profile_path: copy_import_file(user_data[:export]),
```
Layout/FirstHashElementIndentation: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.

diaspora-code-review commented on app/controllers/users_controller.rb:
```diff
@@ -229,15 +233,30 @@ def change_email(user_data)
+  def copy_import_files(user_data)
+    {
+            profile_path: copy_import_file(user_data[:export]),
+            photos_path: copy_import_file(user_data[:exported_photos_file])
```
Layout/HashAlignment: Align the keys and values of a hash literal if they span more than one line.

diaspora-code-review commented on app/controllers/users_controller.rb:
```diff
@@ -229,15 +233,30 @@ def change_email(user_data)
+  end
+
+  def copy_import_file(tmp_file)
+    if tmp_file.present?
```
Style/GuardClause: Use a guard clause (`return unless tmp_file.present?`) instead of wrapping the code inside a conditional expression.

diaspora-code-review commented on app/controllers/users_controller.rb:
```diff
@@ -229,15 +233,30 @@ def change_email(user_data)
+
+  def copy_import_file(tmp_file)
+    if tmp_file.present?
+      file_path_to_save_to = Rails.root.join("public","uploads", "users", "#{current_user.username}_#{tmp_file.original_filename}")
```
- Layout/SpaceAfterComma: Space missing after comma.
- Layout/LineLength: Line is too long. [131/120]

diaspora-code-review commented on app/services/import_service.rb:
```diff
@@ -101,9 +104,13 @@ def create_folder(compressed_file_name)
+  # @param [*String] files
+  def remove_import_files(*files)
+    files.each do |file|
+      if file && File.exist?(file)
```
Style/IfUnlessModifier: Favor modifier `if` usage when having a single-line body. Another good alternative is the usage of control flow `&&`/`||`.

diaspora-code-review commented on app/services/import_service.rb:
```diff
@@ -101,9 +104,13 @@ def create_folder(compressed_file_name)
+  def remove_import_files(*files)
+    files.each do |file|
+      if file && File.exist?(file)
+        File.delete(file) rescue nil
```
Style/RescueModifier: Avoid using `rescue` in its modifier form.

tclaus commented:
I merged latest development into it to fix merge errors. 
This adds a Import UI to the user settings page.

There might be a problem when uploading a json.gz - when unzipped, the resulting file is just a messed up binary. Can someone check this?

cmrd-senya commented:
> There might be a problem when uploading a json.gz - when unzipped, the resulting file is just a messed up binary. Can someone check this?

This may be an issue with the development environment, I've seen something like that happened while downloading the archive on the dev env.

tclaus commented:
> Oh yes, if you download the file in the dev env, it is gzipped twice, because rails again gzips it on download, while in production it will be served by your reverse proxy which doesn't do stuff like that.

Good to be on a team that can help.

cmrd-senya commented on app/services/import_service.rb:
```diff
@@ -3,24 +3,27 @@
 
   def import_by_files(path_to_profile, path_to_photos, username, opts={})
+    user = User.find_by(username: username)
+    raise ArgumentError, "Username #{username} should exist before uploading photos." if user.nil?
```
I think raising exception from the sidekiq job would schedule this job to retry. If this error is irrecoverable we should instead finish the job without errors, possibly logging some error messages.

cmrd-senya commented on app/workers/import_user.rb:
```diff
@@ -4,9 +4,9 @@ module Workers
+    def perform_archive_job(user_id, import_parameters)
       user = User.find(user_id)
-      ImportService.new.import_by_user(user)
+      ImportService.new.import_by_user(user.username, import_parameters)
```
So we have a user object, and then we get it's `username` just to look up the user again inside the `ImportService`. Maybe we can still pass the user object here?

Instead of changing this signature to username, maybe we could change `import_by_files` to accepting user object and in the rake task we could call `import_by_files` passing the `user` object that we find outside of the service?

cmrd-senya commented on lib/archive_importer/post_importer.rb:
```diff
@@ -17,6 +17,7 @@ def substitute_author
 
       entity_data["photos"].each do |photo|
         photo["entity_data"]["author"] = user.diaspora_handle
+        photo["entity_data"]["remote_photo_path"] = "#{AppConfig.pod_uri}uploads\/images\/"
```
Okay, so this is how we link uploaded photos. If there are files in the photos archive that weren't in the end linked to any photo instance, should we detect and clean them up?

cmrd-senya reviewed (commented):
So apart from my comments, a few thoughts:
1. This PR doesn't add new tests, do we want to add tests here, or we don't consider tests mandatory for everything?
2. Do we want to give podmins a configuration option to hide/show import functionality? Is there a reason why a podmin may want to turn off this feature on their pods?
3. Could we combine the two archives -- profile and photos -- in one? We can reduce potential issues if we just needed to upload one file.

tclaus commented on app/controllers/users_controller.rb:
```diff
@@ -122,6 +122,10 @@ def confirm_email
   end
 
+  def import_parameter?(import_parameters)
+    import_parameters[:profile_path] || import_parameters[:photos_path]
```
A user could first import his/her account and in another step the photos. These files are huge, that to separate them still makes sense. 
After the profile is imported the user could import the photos. 
I wouldn't put these files into on (very) big file.

tclaus commented on app/services/import_service.rb:
```diff
@@ -3,24 +3,27 @@
       import_user_photos(user, path_to_photos)
     end
-    remove_file_references(user)
+    remove_import_files(path_to_profile, path_to_photos)
```
What do you mean? The 'import_user_profile' and 'import_user_photos' handle (known) errors. Whenever execution comes here, the uploaded files (I guess the zipped ones) will be removed. 
Could you describe your idea or give a suggestion?

diaspora-code-review commented on app/services/import_service.rb:
```diff
@@ -3,27 +3,35 @@
 
-  private
+  def import_profile_if_present(opts, path_to_photos, path_to_profile, username)
+    if path_to_profile.present?
```
Style/GuardClause: Use a guard clause (`return unless path_to_profile.present?`) instead of wrapping the code inside a conditional expression.

tclaus committed:
- Removed import file from private folder

tclaus commented:
@SuperTux88  FYI:; I have rebased this branch to latest develop and fixed (most, if not, all?) issues. 
May it be a good idea to remove this PR and makle a new one without years old issues? 

This should only a basic UI to start a migration for new users, without any fancy configurable settings.

Flaburgan commented:
Hello @tclaus thank you for coming to this again, with many pods closing we need this feature more than ever. I am not sure we should open a new pull request though. If you fixed the remarks made above, why not simply say it as an answer, as usual?

tclaus commented:
> Hello @tclaus thank you for coming to this again, with many pods closing we need this feature more than ever. I am not sure we should open a new pull request though. If you fixed the remarks made above, why not simply say it as an answer, as usual?

The masses on remarks, thoughts and Linter-Issues makes it not easy to keep an overview. Most real issues should have a "Outdated" label.  

So - what should really, really be done before this is ready to approve?

tclaus committed:
Fix JS issue

denschub commented:
> The masses on remarks, thoughts and Linter-Issues makes it not easy to keep an overview.

This is one of the many reasons why I always ask anyone to please always run linters locally before pushing (in addition to the fact that I don't like being spammed by lots of linting emails).

Given the age of this PR, nobody can remember what was said 4 years ago anyway. If you want, close this PR and open a new one with the same code and it can be reviewed from scratch.

tclaus commented:
Closed. A new PR will be opend with same function, but updated, rebased and without aged comments and issues.
